authz/providers/azure/rbac/checkaccessreqhelper.go

```diff
@@ -0,0 +1,285 @@
+	Description string `json:"description"`
+	Permission
+	Scope                   string `json:"scope"`
+	DoNotApplyToChildScopes bool   `json:"doNotApplyToChildScopes"`
+	principals              []Principal
+	excludeprincipals       []Principal
+	Condition               string `json:"condition"`
+	ConditionVersion        string `json:"conditionVersion"`
+}
+
+type AzureDenyAssignment struct {
+	MetaData          map[string]interface{} `json:"metadata"`
+	IsSystemProtected string                 `json:"isSystemProtected"`
+	IsBuiltIn         bool                   `json:isBuiltIn`
+	DenyAssignment
+}
+
+type AuthorizationDecision struct {
+	Decision            string              `json:"accessDecision"`
+	ActionId            string              `json:"actionId"`
+	IsDataAction        bool                `json:"isDataAction"`
+	AzureRoleAssignment AzureRoleAssignment `json:"roleAssignment"`
+	AzureDenyAssignment AzureDenyAssignment `json:"denyAssignment"`
+	TimeToLiveInMs      int                 `json:"timeToLiveInMs"`
+}
+
+func getScope(resourceId string, attr *authzv1.ResourceAttributes) string {
+	if attr != nil && attr.Namespace != "" {
+		return path.Join(resourceId, namespaces, attr.Namespace)
+	}
+	return resourceId
+}
+
+func isValidUUID(u string) bool {
+	_, err := uuid.Parse(u)
+	return err == nil
+}
+
+func getValidSecurityGroups(groups []string) []string {
+	var finalGroups []string
+	for _, element := range groups {
+		if isValidUUID(element) {
+			finalGroups = append(finalGroups, element)
+		}
+	}
+	return finalGroups
+}
+
+func getActionName(verb string) string {
+	/* special verbs
+	use verb on podsecuritypolicies resources in the policy API group
+	bind and escalate verbs on roles and clusterroles resources in the rbac.authorization.k8s.io API group
+	impersonate verb on users, groups, and serviceaccounts in the core API group
+	userextras in the authentication.k8s.io API group
+
+	https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb
+	*/
+	switch verb {
+	case "get":
+		fallthrough
+	case "list":
+		fallthrough
+	case "watch":
+		return "read"
+
+	case "bind":
+		fallthrough
+	case "escalate":
+		fallthrough
+	case "use":
+		fallthrough
+	case "impersonate":
+		return "action"
+
+	case "create":
+		fallthrough //instead of action create will be mapped to write
+	case "patch":
+		fallthrough
+	case "update":
+		return "write"
+
+	case "delete":
+		fallthrough
+	case "deletecollection": // TODO: verify scenario
+		return "delete"
+	default:
+		return ""
+	}
+}
+
+func getDataAction(subRevReq *authzv1.SubjectAccessReviewSpec, clusterType string) AuthorizationActionInfo {
+	authInfo := AuthorizationActionInfo{
+		IsDataAction: true}
+
+	authInfo.AuthorizationEntity.Id = clusterType
+	if subRevReq.ResourceAttributes != nil {
+		if subRevReq.ResourceAttributes.Group != "" {
+			authInfo.AuthorizationEntity.Id = path.Join(authInfo.AuthorizationEntity.Id, subRevReq.ResourceAttributes.Group)
```

Comment:
why do we append `Group` to `Id`? Can you add a sample Id in comment?

Reply:
Added sample request in beginning of prepareCheckAccessRequestBody.
We add group info in data actions hence adding that as part of Id as below.

"Id": "Microsoft.Kubernetes/connectedClusters/extensions/deployments/read",